docs(mcp): add server

JosephAnson opened the pull request:
### 🔗 Linked issue

<!-- If it resolves an open issue, please link the issue here. For example "Resolves #123" -->
Resolves https://github.com/nuxt/ui/issues/4841

### ❓ Type of change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->

- [x] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

This PR implements a comprehensive MCP server that to provide some useful tools, prompts and resources for the Nuxt UI.

  **Inspiration & Setup:**
  - Inspired by [shadcn-ui-mcp-server](https://github.com/Jpisnice/shadcn-ui-mcp-server)
  - Setup follows the pattern from [mcp-with-nuxt-vercel](https://github.com/atinux/mcp-with-nuxt-vercel)

  **New Features:**
  - MCP server with multiple API endpoints for component data access
  - Component listing and categorization tools
  - Template and migration guide resources
  - Getting started guide management
  - Search functionality for components by category
  - Component example retrieval system

  **Added:**
  - New API endpoints listed under `/server/api/mcp/`
  - MCP transport handler at `/server/routes/mcp/[transport].ts`
  - Documentation page for MCP integration
  
  **Note:** The implementation currently requires a `redisUrl` configuration to function properly. Vercel team assistance may be needed for proper Redis setup in the deployment environment.

### Testing

I recommend if you want to test it locally to use: https://github.com/modelcontextprotocol/inspector

### Demo

<img width="683" height="319" alt="image" src="https://github.com/user-attachments/assets/232e6e4f-030e-48ad-b174-7d55f880fe2b" />

As you can see you can query it to help with adding components, here the ai queries the old version of the component FormGroup (v2) from memory but then corrects itself by querying the mcp server. It's then able to see how it works quickly and provide examples for usage, resulting in a correct refactor.

<img width="1063" height="420" alt="image" src="https://github.com/user-attachments/assets/ff8acf12-3622-496e-8e86-0ffb28ccd06a" />

### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have linked an issue or discussion.
- [x] I have updated the documentation accordingly.

JosephAnson committed:
feat: create mcp server

JosephAnson committed:
chore: typecheck now passes

JosephAnson commented:
I'm currently having issues getting the correct types on the v4 alpha branch, it's not using the correct type for the Nuxt Content queryCollection. It seems like the fix it to add the newer `tsconfig.json` format as this provides the correct types in the IDE, however when running the `typecheck` command it returns errors.

```json
 "references": [
    {
      "path": "./.nuxt/tsconfig.app.json"
    },
    {
      "path": "./.nuxt/tsconfig.server.json"
    },
    {
      "path": "./.nuxt/tsconfig.shared.json"
    },
    {
      "path": "./.nuxt/tsconfig.node.json"
    }
  ],
  // https://nuxt.com/docs/guide/concepts/typescript
  "files": []
```

I've currently marked areas where the code is correct with:
`// @ts-expect-error TODO: This will be fixed when the tsconfig is setup correctly`
You can also see a similar fix in the `docs/server/routes/raw/[...slug].md.get.ts` file.

HugoRCD commented:
@JosephAnson Hi Joseph, sorry I haven't had time to look at the PR in detail yet but from what little I've seen it looks really good! The focus right now is on releasing v4 of `@nuxt/ui`. So once it's released, I'll come and have a look at it in detail and lend a hand, because it was one of our focuses after v4 to make an MCP server!

JosephAnson commented:
Great news, sounds good to me, thanks @HugoRCD. For now I'll keep using it locally unless it's possible to have a test url related to this PR? I'll keep testing and making tweaks if anything can be improved.

HugoRCD commented:
> Great news, sounds good to me, thanks @HugoRCD. For now I'll keep using it locally unless it's possible to have a test url related to this PR? I'll keep testing and making tweaks if anything can be improved.

Do you have good local results?

JosephAnson commented:
Locally it's working well, it's not necessary for a deployed dev url for it work, however it would save me running the dev server each time haha. Just trying to be lazy with it. Only problem I could see is you'll need to run the same version of Nuxt UI for it work correctly, as it'll always suggest the changes of the version it's deployed with.

HugoRCD commented:
> Locally it's working well, it's not necessary for a deployed dev url for it work, however it would save me running the dev server each time haha. Just trying to be lazy with it. Only problem I could see is you'll need to run the same version of Nuxt UI for it work correctly, as it'll always suggest the changes of the version it's deployed with.

I was talking more about the quality of the ia's responses and whether there are still many hallucinations with it or not!

JosephAnson commented:
It's working great tbh, code refactors have been much better. Before generally it would try to use old versions of the components, wouldn't understand which v-models to bind, now it's checking directly. With the component meta lookup it's been pretty great, my initial plan was something external but haven't that additional info seems invaluable. 

Additionally things like having upgrading to v4 for you automatically will work well once the migration guide is populated correctly, as there's a tool for that using the markdown helper used for the /raw endpoint.

I've attached some demo screenshots to the pr description for reference.

JosephAnson committed:
refactor: drop the mcp-handler dependency

benjamincanac committed:
Merge branch 'v4' into pr/4878

benjamincanac committed:
docs(mcp): move into ai dir

benjamincanac commented on docs/server/api/list-examples.get.ts:
```diff
@@ -0,0 +1,11 @@
+import { appendHeader, defineEventHandler } from 'h3'
```
Why is this route not in the `mcp/` directory? 🤔

JosephAnson replied:
I wasn't sure where to put it as /api/component-example was already outside and /api/list-examples is a match with that endpoint. Maybe /list-component-examples or /component-examples would've been a better name also.

benjamincanac replied:
Yes but this `list-examples` endpoint is only used inside the `mcp` right? 🤔

JosephAnson replied:
Yeah it's true, it's only used by the mcp, I was caught in two minds, I wasn't sure what you'd prefer. We can move it inside the mcp folder.

JosephAnson commented:
I'd like to drop the `mcp-handler` and `redis` dependency as the approach is outdated using SSE and instead use Streamable HTTP (see here: https://modelcontextprotocol.io/specification/2025-06-18/basic/transports#streamable-http, following the guide here: https://github.com/modelcontextprotocol/typescript-sdk#without-session-management-stateless), it's quicker as you no longer need to use `npx`, it feels more responsive too. I have a commit ready, i'd like to push the change or would you rather me create a PR for this PR?

benjamincanac committed:
docs(mcp): rename title

benjamincanac commented:
@JosephAnson Even better without the `redis` dependency 🎉

benjamincanac commented on docs/content/docs/1.getting-started/7.ai/1.mcp.md:
```diff
@@ -0,0 +1,79 @@
+---
```
@JosephAnson Would it be possible to have in the docs a section for each editor (Cursor, Windsurf, etc.) like on https://chakra-ui.com/docs/get-started/ai/mcp-server?

It would match the `llms.txt` page as well.

JosephAnson replied:
I've updated the documentation to follow a similar format

HugoRCD committed:
use `getValidatedQuery`

HugoRCD commented on docs/server/routes/mcp/[transport].ts:
```diff
@@ -0,0 +1,320 @@
+  })
+
+  event.node.res.on('close', () => {
+    console.log('Request closed')
```
Is it really necessary to log every request? 🤔

JosephAnson replied:
No it's not, that can be removed.

HugoRCD commented:
@JosephAnson I don't know if I've misconfigured my Cursor or not but I'm getting this error, I think since the change to `streamable-HTTP` 🤔

<img width="818" height="1268" alt="CleanShot 2025-09-08 at 16 33 20@2x" src="https://github.com/user-attachments/assets/ece5468c-88b2-404a-8b66-b6561b288d64" />

JosephAnson committed:
chore: remove console.log

JosephAnson committed:
refactor: move list-examples endpoint inside /mcp

JosephAnson commented:
> @JosephAnson I don't know if I've misconfigured my Cursor or not but I'm getting this error, I think since the change to `streamable-HTTP` 🤔
> 
> <img alt="CleanShot 2025-09-08 at 16 33 20@2x" width="818" height="1268" src="https://private-user-images.githubusercontent.com/71938701/486848583-ece5468c-88b2-404a-8b66-b6561b288d64.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTczNDU5ODYsIm5iZiI6MTc1NzM0NTY4NiwicGF0aCI6Ii83MTkzODcwMS80ODY4NDg1ODMtZWNlNTQ2OGMtODhiMi00MDRhLThiNjYtYjY1NjFiMjg4ZDY0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTA4VDE1MzQ0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY3ZDg3OGE5OWU5YWY5ZDhhMjg5ZTJmNDk2NDUwZjUyYmFjNDgyMjY0NDc3YmUzZjAwNjI1NzgyYmY5ZjcxMmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GREAv7455JvbHnCUPM7CAESUp13hLrjY2QUn9g0aH9g">

You need to modify mcp server config to:

```
 "nuxt-ui": {
      "type": "http",
      "url": "https://ui4.nuxt.com/mcp/sse"
    }
```

I updated the docs in the previous commit. Let me know if it's working for you after that change.

For claude I use the command `claude mcp add --transport http nuxt-ui https://localhost:3000/mcp/sse`

HugoRCD commented:
> > @JosephAnson I don't know if I've misconfigured my Cursor or not but I'm getting this error, I think since the change to `streamable-HTTP` 🤔
> > <img alt="CleanShot 2025-09-08 at 16 33 20@2x" width="818" height="1268" src="https://private-user-images.githubusercontent.com/71938701/486848583-ece5468c-88b2-404a-8b66-b6561b288d64.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTczNDU5ODYsIm5iZiI6MTc1NzM0NTY4NiwicGF0aCI6Ii83MTkzODcwMS80ODY4NDg1ODMtZWNlNTQ2OGMtODhiMi00MDRhLThiNjYtYjY1NjFiMjg4ZDY0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTA4VDE1MzQ0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY3ZDg3OGE5OWU5YWY5ZDhhMjg5ZTJmNDk2NDUwZjUyYmFjNDgyMjY0NDc3YmUzZjAwNjI1NzgyYmY5ZjcxMmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GREAv7455JvbHnCUPM7CAESUp13hLrjY2QUn9g0aH9g">
> 
> You need to modify mcp server config to:
> 
> ```
>  "nuxt-ui": {
>       "type": "http",
>       "url": "https://ui4.nuxt.com/mcp/sse"
>     }
> ```
> 
> I updated the docs in the previous commit. Let me know if it's working for you after that change

This was all that was missing, it works fine now 😁

JosephAnson commented:
It should be possible for this to work on the preview url now, however there seems to be some server errors when visiting https://ui-4194m0mel-nuxt-js.vercel.app/mcp/sse. I think you'll need to get the server logs to debug what's happening. All the new mcp apis are working correctly when testing them manually.

HugoRCD commented:
> It should be possible for this to work on the preview url now, however there seems to be some server errors when visiting [ui-4194m0mel-nuxt-js.vercel.app/mcp/sse](https://ui-4194m0mel-nuxt-js.vercel.app/mcp/sse). I think you'll need to get the server logs to debug what's happening. All the new mcp apis are working correctly when testing them manually.

This is the server error:
<img width="1504" height="418" alt="CleanShot 2025-09-08 at 17 03 14@2x" src="https://github.com/user-attachments/assets/fbf90fcc-86be-40b2-8d5d-bbd7393d2668" />

HugoRCD commented on docs/server/api/mcp/list-examples.get.ts:
Maybe remove the `.get` and put it in `defineCachedEventHandler` no?

HugoRCD committed:
up

JosephAnson committed:
refactor: use the v3 version of zod 4 and ignore type issues as it's working correctly

Later we can update the new version of modelcontextprotocol/sdk once it's released and resolve the type issues

JosephAnson committed:
chore: update mcp documentation

benjamincanac commented on docs/server/routes/mcp/[transport].ts:
```diff
@@ -0,0 +1,324 @@
+  return server
+}
+
+// Create MCP handler using Vercel adapter
```
Can we remove this comment?
```suggestion
```

benjamincanac commented on docs/server/routes/mcp/[transport].ts:
```diff
@@ -0,0 +1,324 @@
+}
+
+// Create MCP handler using Vercel adapter
+export default defineEventHandler(async (event) => {
```
Does it make sense to have this route on `/mcp/sse` since it's now using HTTP? 🤔

atinux replied:
I confirm that `/mcp` would be perfect!

atinux replied:
Another suggestion, is to check at first if the request header "accept" includes "text/html", then we can redirect to the docs page about the MCP server

JosephAnson replied:
Yeah good idea, I've move it to /mcp and based on the header like you suggested I've redirected to the documentation page.

atinux commented:
Awesome work @JosephAnson 💚

benjamincanac commented:
@JosephAnson I had a project a while back to implement AI search in the docs (https://github.com/nuxt/ui/pull/4114/files) but couldn't make it work with a RAG. Could this new MCP you implemented help us achieve this?

JosephAnson committed:
refactor: move mcp server to /mcp endpoint instead of /mcp/sse

JosephAnson committed:
Merge branch 'v4' into feat/mcp-server-v4

* v4:
  perf(module): do not block setup by importing plugin (#4923)
  docs(nuxt.config): add vercel runtime cache config (#4930)

JosephAnson commented:
> @JosephAnson I had a project a while back to implement AI search in the docs ([#4114 (files)](https://github.com/nuxt/ui/pull/4114/files)) but couldn't make it work with a RAG. Could this new MCP you implemented help us achieve this?

Tbh I'm not sure, it seems like what you've made should work provided that the embeddings are in the db, you can add the mcp tools to the `AI` streamText, that could help it make a better conclusion, I'll take a look this week. Not sure about the database instance now though, if you've moved hosting to vercel as https://hub.nuxt.com/docs/features/autorag seems to be only supported on cloudflare. You will need a storage option from vercel to store the embeddings.

benjamincanac committed:
Merge branch 'v4' into feat/mcp-server-v4

JosephAnson committed:
feat: add list_documentation_pages tool to correctly identify the paths for the get_documentation_page tool

JosephAnson commented:
Sorry for a last minute change, added a new tool called `list_documentation_pages` to list all the correct paths in the documentation as I noticed when `get_content_page` was called it would try an incorrect path sometimes, this would provide a solution for the ai to get the correct paths. Also renamed `get_content_page` to `get_documentation_page` as I feel that's more accurate.

benjamincanac commented:
Sure! Let me know when you think this is ready 😊

Doesn't this make the `list_getting_started_guides` tool useless now?

JosephAnson commented:
> Sure! Let me know when you think this is ready 😊
> 
> Doesn't this make the `list_getting_started_guides` tool useless now?

I think it's ready now, at least enough to move ahead with as v1. That's a good point, I think it still serves a purpose as it's a direct access for a certain use case, same as the `get_migration_guide`. Although if you'd prefer we could remove it for now?

benjamincanac commented:
No we can keep it I was just wondering, maybe what we're missing is a `list_composables` then? 🤔

JosephAnson commented:
Yeah that's not a bad idea to add another tool for that, generally it's recommended to keep it as atomic as possible. So I think it's not a bad idea. I can also add it as a resource as well.

JosephAnson committed:
feat: add list composables tool and add documentation pages to mcp resources

JosephAnson committed:
chore: clean up endpoints that were returning unused data

JosephAnson commented:
Ok I think we're all good to go, unless you see some issues that need fixing i'm happy to consider it ready. Only issue I could see is currently `useFormField` is returned by the `list_composables` tool and currently the documentation for it is set to `navigation: false`, should we also hide it from the results? It is possible to navigate to it via https://ui4.nuxt.com/docs/components/form#input-events however.

<img width="615" height="355" alt="image" src="https://github.com/user-attachments/assets/c574461b-84af-40d2-8d68-f70d4da2ac20" />

benjamincanac commented on docs/server/api/mcp/list-components.ts:
```diff
@@ -0,0 +1,19 @@
+    .all()
+
+  // Extract component names from paths
+  return componentDocs
```
Why not call it `components` here and use the `title`? 🤔

JosephAnson replied:
Good catch, this was left over from when i was first calling the `get_component` tool with the kebab-case version, I was getting it from the path as there was an edge case with `PageCTA` before I was manually converting from PageCTA to `page-c-t-a` which wasn't a match as the path is `/page-cta`, now with using the `scule` library it doesn't suffer from the same issue. This could lead to issues in the future as we're hoping the conversion from name to path will always work.

I'm now passing back the whole object array, it'll work better with `get_documentation_page` too, as the path is available for both the `list_composables` and `list_components`.

JosephAnson committed:
refactor: provide more information from the components and composable tools to also synergize with the `get_documentation_page`

benjamincanac commented:
Regarding the `useFormField ` composable we can leave it like this I guess, we'll have to document this page at some point anyway 😅

benjamincanac commented:
Thanks again @JosephAnson for the awesome work 💚

sandros94 commented:
Somehow I completely missed this PR 😅

Just wanted to leave a couple of notes I've discovered while building MCP servers with various models, might come handy in future iterations:

- hallucinations come from a out-of-memory context issue, essentially how many tokens are required for the current inference. Which are usually distributed as (in order of higher token consumption and sooner hallucination)
  1. manually added resources (either via mcp server or local client)
  2. data returned by tools listing and calls
  3. conversation length (tho this can easily become no. 1 or 2)
  4. times an LLM need to call a tool (either local or remote)
  5. system and user prompts
- AI search (aka vector/semantic search) requires:
  - source data to be embedded and kept track for future embedding, like having an hash of the current doc page and the various examples shown in it (re-embedding everything on each build is just a waste of resources)
  - embedding the request that is coming from the user's search, then doing a vector search against our data and providing it to the user

I've solved the first one by creating detailed instructions (based on the project) and a small parser of the UI docs to extract relevant information (using the usage, examples, api and theme chapters we have in all components pages) and informing the LLM that it can in a single tool request get one or multiple chapters for one or multiple components. Here is an [example of the tool](https://github.com/sandros94/demo-h3-mcp/blob/7d4ac262186cfba768b4069522147c39c585d136/src/main.ts#L43-L66) I regularly use (both name, description and descriptions in the jsonSchema are equally important to let the LLM understand what to bet and what to expect), while [here is the Nuxt UI doc parser](https://github.com/sandros94/demo-h3-mcp/blob/7d4ac262186cfba768b4069522147c39c585d136/src/utils/docs.ts#L142-L164) to extract only relevant part (either by letting the LLM decide or directly asking to include/exclude something)

While for AI search it is something that I've stopped testing for now, as I did reach most of what I was looking for via MCP tools. I'm starting to think that a vector search is not that much useful in a docs website, or at least not when you have to factor in that most of the meaning (based on vector size) might be constructed from common symbols and common css classes throughout the documentation. But I might just be wrong.

HugoRCD commented:
@JosephAnson Do you have any links to your socials to share with us for the server release post?

JosephAnson commented:
> @JosephAnson Do you have any links to your socials to share with us for the server release post?

Not sure which one you'd like to share, you can pick what you prefer: 

Github: https://github.com/JosephAnson
Website: http://www.josephanson.com/
Linkedin: https://www.linkedin.com/in/josephanson/
